river/command.zig

```diff
@@ -117,3 +117,71 @@ pub fn errToMsg(err: Error) [:0]const u8 {
         Error.Other => unreachable,
     };
 }
+
+
+/// args[0] must be the first option if any
+/// offset will be set to the index of the first entry in args that is not an option
+/// Return an error when parsing a value for an option fails
+/// TODO where to enforce the naming of arguments
+fn parse(args: []const []const u8, offset: *usize, comptime T: type) !T {
+    if (@typeInfo(T) != .Struct) @compileError("T must be a struct");
+
+    // The empty {} make sure there is a default value for all members
+    var ret: T = T{};
+    // Make sure all bools are false by default
+    inline for (@typeInfo(T).Struct.fields) |field| {
+        if (@typeInfo(field.field_type) != .Bool) continue;
+        if(field.default_value.?) @compileError("bool options must be false by default");
+
+    }
+    // TODO cleaner way to set offset
+    var i: usize = 0;
+    defer offset.* = i;
+
+    outer: for (args) |arg, idx| {
+        i = idx + 1;
+        var found = false;
+        inline for (@typeInfo(T).Struct.fields) |field| {
+            if (std.mem.eql(u8, arg, field.name)) {
+                found = true;
+                switch (@typeInfo(field.field_type)) {
+                    .Bool => {
+                        @field(ret, field.name) = true;
+                    },
+                    .Int => {
+                        if (args.len < idx + 2) return error.MissingArgument;
+                        @field(ret, field.name) = try std.fmt.parseInt(field.field_type, args[idx+1], 10);
+                        continue :outer;
```

Comment:
This is insufficient as the next iteration must be skipped as well, but it doesn't matter as we don't need to parse ints yet :)

Reply:
Yeah. Removed it.